Automattic/woocommerce-payments
# Add HMR and split webpack configuration

#### Changes proposed in this Pull Request
- Update `@wordpress/dependency-extraction-webpack-plugin` to support react refresh extraction.
- Installed the following required dependencies:
  - `@pmmmwh/react-refresh-webpack-plugin@0.5.7` – To load the runtime on our builds.
  - `react-refres@0.14.0` - The actual React Refresh code, used for Babel.
  - `webpack-dev-server@4.9.3` – Required to serve HMR changes.
  - `webpack-merge@5.8.0` – Used to easily merge different configurations effectively overriding some rules.
- Split webpack configuration into `shared`, `production`, and `development`– Adding HMR requires a bunch of settings that will clutter our production one, having them separated will make updates easier.
- Add HMR webpack configuration conditionally based on `WEBPACL_SERVE` from `env`.
- Load runtime script for admin area. As single runtime is required to make HMR work while using multiple entries on the same page. It will only load if `SCRIPT_DEBUG` is enabled and the runtime script exists.
- Add `npm run hmr` command with instructions.

I'd like to highlight the need for the runtime script and how it's loaded from the PHP side. Ideally we could get rid of it joining the runtime and a vendor chunk together, but for that we will need to properly configure split chunks and try to slim down or bundled dependencies. It's something more suited for a bundle optimization PR, where we could greatly reduce our bundle size.

#### Demo
https://user-images.githubusercontent.com/7670276/182116957-7e027bec-cc9d-456c-8d3d-98633bc45c99.mov

#### Testing instructions
The current implementation relies on WP loaded React Refresh, so it requires at least WordPress 6.

- Run `npm run build:release` before and after applying this branch.
- Compare the output, apart from the hashes everything else should match.
- Run `npm run hmr`.
- Navigate to a React page like WCPay settings or Multi-Currency.
- Update something from a mounted component.
- It should update without refreshing the page.

-------------------


**Post merge**
